src/lib.rs

```diff
@@ -1,748 +1,268 @@
-    // desc_1_0: IRRootSignatureDescriptor,
-    pub desc_1_1: IRRootSignatureDescriptor1,
-}
-
-#[repr(u32)]
-#[derive(Copy, Clone)]
-pub enum IRRootSignatureVersion {
-    IRRootSignatureVersion_1_0 = 0x1,
-    IRRootSignatureVersion_1_1 = 0x2,
-}
-
-#[repr(C)]
-#[derive(Copy, Clone)]
-pub struct IRVersionedRootSignatureDescriptor {
-    pub version: IRRootSignatureVersion,
-    pub u: IRVersionedRootSignatureDescriptor_u,
-}
-
-#[derive(Debug)]
-struct IRRootSignatureFn<'lib> {
-    create_from_descriptor: libloading::Symbol<
-        'lib,
-        unsafe extern "C" fn(
-            *const IRVersionedRootSignatureDescriptor,
-            *mut *mut IRErrorOpaque,
-        ) -> *mut IRRootSignatureOpaque,
-    >,
-    destroy: libloading::Symbol<'lib, unsafe extern "C" fn(*const IRRootSignatureOpaque) -> ()>,
-    get_resource_count: libloading::Symbol<'lib, unsafe extern "C" fn() -> ()>,
-    get_resource_locations: libloading::Symbol<'lib, unsafe extern "C" fn() -> ()>,
-}
-
-pub struct IRRootSignature<'lib> {
-    me: *mut IRRootSignatureOpaque,
-    funcs: IRRootSignatureFn<'lib>,
-}
-
-impl<'lib> Drop for IRRootSignature<'lib> {
+impl Drop for IRRootSignature {
     fn drop(&mut self) {
-        unsafe { (self.funcs.destroy)(self.me) }
+        unsafe { self.funcs.IRRootSignatureDestroy(self.me) }
     }
 }
 
-impl<'lib> IRRootSignature<'lib> {
+impl IRRootSignature {
     pub fn create_from_descriptor(
-        lib: &'lib libloading::Library,
+        compiler: &IRCompiler,
         desc: &IRVersionedRootSignatureDescriptor,
-    ) -> Result<IRRootSignature<'lib>, Box<dyn std::error::Error>> {
+    ) -> Result<IRRootSignature, Box<dyn std::error::Error>> {
         unsafe {
-            let funcs = IRRootSignatureFn {
-                create_from_descriptor: lib.get(b"IRRootSignatureCreateFromDescriptor")?,
-                destroy: lib.get(b"IRRootSignatureDestroy")?,
-                get_resource_count: lib.get(b"IRRootSignatureGetResourceCount")?,
-                get_resource_locations: lib.get(b"IRRootSignatureGetResourceLocations")?,
-            };
+            let mut error = std::ptr::null_mut();
 
-            let mut error: *mut IRErrorOpaque =
-                unsafe { std::ptr::null_mut::<IRErrorOpaque>().add(0) };
+            let me = compiler
+                .funcs
+                .IRRootSignatureCreateFromDescriptor(desc, &mut error);
 
-            let me = (funcs.create_from_descriptor)(desc, &mut error);
+            Ok(Self {
+                me,
+                funcs: compiler.funcs.clone(),
+            })
+        }
+    }
+}
 
-            dbg!(error);
+/// [IRCompilerFactory] is used to load the metal_irconverter dynamic library and holds its functions in an Arc.
+/// Since [IRCompiler] is not thread-safe, this struct provides an interface to create [IRCompiler] instances.
+/// This way, the library only has to be loaded once, but each thread can have its own [IRCompiler] instance.
```

Comment:
```suggestion
/// [`IRCompilerFactory`] is used to load the metal_irconverter dynamic library and holds its functions in an Arc.
/// Since [`IRCompiler`] is not thread-safe, this struct provides an interface to create [`IRCompiler`] instances.
/// This way, the library only has to be loaded once, but each thread can have its own [`IRCompiler`] instance.
```